vllm/v1/attention/backends/utils.py

```diff
@@ -679,13 +750,82 @@ def subclass_attention_metadata(
     return Wrapped
 
 
+@functools.lru_cache
 def make_kv_sharing_fast_prefill_attention_metadata(
-    metadata_cls: Any, ) -> Any:
+    metadata_cls: Hashable, ) -> Any:
     """
     Return a new subclass of `metadata_cls` for fast prefill
     """
-    return subclass_attention_metadata(
+    attn_metadata_dataclass = subclass_attention_metadata(
         name_prefix="KVSharingFastPrefill",
         metadata_cls=metadata_cls,
         fields=KV_SHARING_FAST_PREFILL_METADATA_FIELDS,
     )
+    # Make attention metadata type inherit
+    # KVSharingFastPrefillAttentionMetadata type
+    fast_prefill_metadata_type = type(
+        attn_metadata_dataclass.__name__,
+        (
+            attn_metadata_dataclass,
+            KVSharingFastPrefillAttentionMetadata,
+        ),
+        {},
+    )
+    return fast_prefill_metadata_type
+
+
+@runtime_checkable
+class KVSharingFastPrefillAttentionMetadata(Protocol):
+    logits_indices_padded: torch.Tensor
+    num_logits_indices: int
+
+
+def create_kv_sharing_fast_prefill_attn_metadata_subclass(
+    metadata: Any,
+    common_attn_metadata: CommonAttentionMetadata,
+) -> Any:
+    # Dynamically create a a dataclass type that inherits
```

Comment:
can we change it to something similar to how we create the `FastPrefillAttentionBuilder`
```
    class KVSharingFastPrefillAttentionMetadata(metadata.__class__, KVSharingFastPrefillMetadata):
         def __init__(self, metadata, common_attn_metadata):
                ....
    return KVSharingFastPrefillAttentionMetadata(metadata, common_attn_metadata)
```
where `KVSharingFastPrefillMetadata` is the `KVSharingFastPrefillAttentionMetadata` you defined